Sources/SourceControl/GitRepository.swift

```diff
@@ -207,12 +210,21 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
 
     public func isValidDirectory(_ directory: Basics.AbsolutePath) throws -> Bool {
         let result = try self.git.run(["-C", directory.pathString, "rev-parse", "--git-dir"])
-        return result == ".git" || result == "." || result == directory.pathString
+        return result == ".git" || result == "."
+            // Compare the canonical representation, which will drop any suffix
+            || CanonicalPackageLocation(result) == CanonicalPackageLocation(directory.pathString)
     }
 
     public func isValidDirectory(_ directory: Basics.AbsolutePath, for repository: RepositorySpecifier) throws -> Bool {
         let remoteURL = try self.git.run(["-C", directory.pathString, "config", "--get", "remote.origin.url"])
-        return remoteURL == repository.url
+        switch repository.location {
+        case .url(let url):
+            // Compare the canonical representation, which will drop any suffix and canonicalize scp-style urls
+            return CanonicalPackageURL(remoteURL) == CanonicalPackageURL(url.absoluteString)
+        case .path(let absolutePath):
+            // Compare the canonical representation, which will drop any suffix
+            return CanonicalPackageLocation(remoteURL) == CanonicalPackageLocation(absolutePath.pathString)
```

Comment:
I doubt this matters in practice, but the [caching logic](https://github.com/swiftlang/swift-package-manager/blob/main/Sources/SourceControl/RepositoryManager.swift#L617) always uses URL and it's probably better to do that here as well. That wouldn't fix the case mentioned in the description, but to fix that you could update `computeCanonicalLocation` to return a `file` scheme if one wasn't given (which seems somewhat reasonable given we're already inferring `ssh` if the user is `git`). Does that seem fine to you @neonichu?